source-sftp/main.go

```diff
@@ -0,0 +1,397 @@
+		if len(l.stack) == 0 {
+			// No more files to list. io.EOF is the expected error in this case.
+			return filesource.ObjectInfo{}, io.EOF
+		}
+
+		// Pop the next file in lexical order.
+		next := l.stack[len(l.stack)-1]
+		l.stack = l.stack[:len(l.stack)-1]
+
+		// If this file is a symlink, it might be a symlink to a directory. We don't descend into
+		// directories so they should not be included in any listing output. If the file is a
+		// symlink to a non-directory, we should report the info of the referent file.
+		if next.Mode()&os.ModeSymlink != 0 {
+			info, err := l.client.Stat(next.path)
+			if err != nil {
+				return filesource.ObjectInfo{}, fmt.Errorf("stat'ing symbolic link: %w", err)
+			}
+			if info.IsDir() {
+				logrus.WithFields(logrus.Fields{
+					"Name": info.Name(),
+					"Mode": info.Mode().String(),
+				}).Info("skipping symbolic link to directory in listing output")
+				continue
+			}
+
+			next = visit{FileInfo: info, path: next.path}
+		}
+
+		if next.IsDir() {
+			// If the query is recursive, descend into this subdirectory to list its files but do
+			// not return info the subdirectory file entry itself.
+			if l.recursive {
+				if err := l.pushDir(next.path); err != nil {
+					return filesource.ObjectInfo{}, fmt.Errorf("pushing next path %s: %w", next.path, err)
+				}
+				continue
+			}
+
+			// If the query is not recursive, do not descend into this subdirectory but do return
+			// info with IsPrefix=true for it. This is accomplished the the next.IsDir() call in the
+			// struct below.
+		}
+
+		return filesource.ObjectInfo{
+			Path:            next.path,
+			IsPrefix:        next.IsDir(),
+			ContentSum:      "",
+			Size:            next.Size(),
+			ContentType:     "",
+			ContentEncoding: "",
+			ModTime:         next.ModTime(),
+		}, nil
+	}
+}
+
+func (l *sftpListing) pushDir(dir string) error {
+	infos, err := l.fs.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	// Filter out file names < query.StartAt and map to visits.
+	visits := []visit{}
+	for _, info := range infos {
+		filePath := sftp.Join(dir, info.Name())
+		if filePath >= l.startAt {
+			visits = append(visits, visit{FileInfo: info, path: filePath})
+		}
+	}
+
+	// Ensure files are lexically sorted in reverse order so earliest files are popped first. We
+	// make no assumption about lexical ordering of files returned from ReadDir etc. as this
+	// ordering is filesystem dependent. The filesource framework requires listings to be in lexical
+	// order, so an explicit sort is required.
+	sort.Slice(visits, func(i, j int) bool {
+		return visits[i].path > visits[j].path
+	})
+
+	l.stack = append(l.stack, visits...)
+
+	return nil
+}
+
+// visit wraps fs.FileInfo and retains the full path to the file, since fs.FileInfo only provides
+// the file name.
+type visit struct {
+	fs.FileInfo
+	path string
+}
+
+func (s *sftpSource) Read(_ context.Context, obj filesource.ObjectInfo) (io.ReadCloser, filesource.ObjectInfo, error) {
```

Comment:
Yeah in practice the runtime will nuke any connectors it needs to and there is no strict requirement for connectors to monitor/propagate signals or context cancellations, although it would be nicer if the `sftp` methods could. But it shouldn't be a problem.